File: PROJET-INTEGRE/microservice_service/src/main/java/com/example/microservice_service/service/OrderService.java
package com.example.microservice_service.service;

import com.example.microservice_service.client.UserClient;
import com.example.microservice_service.dto.UserResponse;
import com.example.microservice_service.entity.FreelancerService;
import com.example.microservice_service.entity.Order;
import com.example.microservice_service.entity.enums.OrderStatus;
import com.example.microservice_service.repository.OrderRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class OrderService {

    private final OrderRepository orderRepository;
    private final FreelancerServiceService freelancerServiceService;
    private final UserClient userClient;

    public Order createOrder(Long buyerId, Long serviceId, String selectedAddOns, java.math.BigDecimal totalPrice) {
        // Validate buyer exists and is active (both CLIENT and FREELANCER can order)
        UserResponse buyer = userClient.getUserById(buyerId);
        if (!buyer.isVerified()) {
            throw new IllegalStateException("User must be verified to place an order.");
        }
        if (!buyer.isActive()) {
            throw new IllegalStateException("User account is disabled.");
        }

        FreelancerService service = freelancerServiceService.getServiceById(serviceId);

        // Cannot order your own service
        if (service.getShop().getFreelancerId().equals(buyerId)) {
            throw new IllegalStateException("You cannot order your own service.");
        }

        Order order = new Order();
        order.setBuyerId(buyerId);
        order.setSellerId(service.getShop().getFreelancerId());
        order.setService(service);
        order.setSelectedAddOns(selectedAddOns);
        order.setTotalPrice(totalPrice);
        order.setStatus(OrderStatus.PENDING_REQUIREMENTS);
        order.setRevisionsUsed(0);

        return orderRepository.save(order);
    }

    public Order getOrderById(Long id) {
        return orderRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Order not found with id: " + id));
    }

    public List<Order> getOrdersByBuyer(Long buyerId) {
        return orderRepository.findByBuyerId(buyerId);
    }

    public List<Order> getOrdersBySeller(Long sellerId) {
        return orderRepository.findBySellerId(sellerId);
    }

    public List<Order> getOrdersByService(Long serviceId) {
        return orderRepository.findByServiceId(serviceId);
    }

    public Order submitRequirements(Long orderId, String requirementsAnswer, Long requesterId) {
        Order order = getOrderById(orderId);
        if (!order.getBuyerId().equals(requesterId)) {
            throw new IllegalStateException("Only the buyer can submit requirements.");
        }
        if (order.getStatus() != OrderStatus.PENDING_REQUIREMENTS) {
            throw new IllegalStateException("Requirements already submitted.");
        }
        order.setBuyerRequirementsAnswer(requirementsAnswer);
        order.setStatus(OrderStatus.IN_PROGRESS);
        int deliveryDays = order.getService().getDeliveryDays();
        order.setDeadline(LocalDateTime.now().plusDays(deliveryDays));
        return orderRepository.save(order);
    }

    public Order deliverOrder(Long orderId, String deliveryMessage, String deliveryFileUrls, Long requesterId) {
        Order order = getOrderById(orderId);
        if (!order.getSellerId().equals(requesterId)) {
            throw new IllegalStateException("Only the seller can deliver the order.");
        }
        if (order.getStatus() != OrderStatus.IN_PROGRESS && order.getStatus() != OrderStatus.REVISION_REQUESTED) {
            throw new IllegalStateException("Order is not in a deliverable state.");
        }
        order.setDeliveryMessage(deliveryMessage);
        order.setDeliveryFileUrls(deliveryFileUrls);
        order.setStatus(OrderStatus.DELIVERED);
        return orderRepository.save(order);
    }

    public Order requestRevision(Long orderId, String revisionNotes, Long requesterId) {
        Order order = getOrderById(orderId);
        if (!order.getBuyerId().equals(requesterId)) {
            throw new IllegalStateException("Only the buyer can request a revision.");
        }
        if (order.getStatus() != OrderStatus.DELIVERED) {
            throw new IllegalStateException("Can only request revision on a delivered order.");
        }
        if (order.getRevisionsUsed() >= order.getService().getRevisionCount()) {
            throw new IllegalStateException("Maximum revisions reached.");
        }
        order.setRevisionNotes(revisionNotes);
        order.setRevisionsUsed(order.getRevisionsUsed() + 1);
        order.setStatus(OrderStatus.REVISION_REQUESTED);
        return orderRepository.save(order);
    }

    public Order completeOrder(Long orderId, Long requesterId) {
        Order order = getOrderById(orderId);
        if (!order.getBuyerId().equals(requesterId)) {
            throw new IllegalStateException("Only the buyer can complete the order.");
        }
        if (order.getStatus() != OrderStatus.DELIVERED) {
            throw new IllegalStateException("Can only complete a delivered order.");
        }
        order.setStatus(OrderStatus.COMPLETED);
        order.setCompletedAt(LocalDateTime.now());
        return orderRepository.save(order);
    }

    public Order cancelOrder(Long orderId, Long requesterId) {
        Order order = getOrderById(orderId);
        if (!order.getBuyerId().equals(requesterId) && !order.getSellerId().equals(requesterId)) {
            throw new IllegalStateException("Only buyer or seller can cancel the order.");
        }
        if (order.getStatus() == OrderStatus.COMPLETED) {
            throw new IllegalStateException("Cannot cancel a completed order.");
        }
        order.setStatus(OrderStatus.CANCELLED);
        return orderRepository.save(order);
    }

    public Order disputeOrder(Long orderId, Long requesterId) {
        Order order = getOrderById(orderId);
        if (!order.getBuyerId().equals(requesterId) && !order.getSellerId().equals(requesterId)) {
            throw new IllegalStateException("Only buyer or seller can dispute the order.");
        }
        if (order.getStatus() == OrderStatus.COMPLETED || order.getStatus() == OrderStatus.CANCELLED) {
            throw new IllegalStateException("Cannot dispute a completed or cancelled order.");
        }
        order.setStatus(OrderStatus.DISPUTED);
        return orderRepository.save(order);
    }
}

File: Esprit-pi-4sae2-2026-freelanceplatform-feature-integrated-frontend/src/app/front/services/order.service.ts
// order.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AuthService } from './auth.service';

export interface Order {
  id: number;
  orderNumber: string;
  buyerId: number;
  sellerId: number;
  serviceId: number;
  service?: any;
  selectedAddOns?: string;
  totalPrice: number;
  status: OrderStatus;
  requirements?: string;
  requirementsAnswer?: string;
  deliveryMessage?: string;
  deliveryFileUrls?: string;
  revisionNotes?: string;
  revisionCount: number;
  maxRevisions: number;
  createdAt: string;
  updatedAt: string;
  deliveredAt?: string;
  completedAt?: string;
}

export type OrderStatus = 
  | 'PENDING_ACCEPTANCE'
  | 'ACCEPTED'
  | 'IN_PROGRESS'
  | 'DELIVERED'
  | 'IN_REVISION'
  | 'COMPLETED'
  | 'CANCELLED'
  | 'DISPUTED';

export interface CreateOrderRequest {
  serviceId: number;
  selectedAddOns?: string;
  totalPrice: number;
  requirements?: string;
}

@Injectable({ providedIn: 'root' })
export class OrderService {
  private readonly BASE_URL = 'http://localhost:8085/microservice-service/api/orders';

  constructor(
    private http: HttpClient,
    private authService: AuthService
  ) {}

  private authHeaders(): HttpHeaders {
    const token = this.authService.getToken();
    return new HttpHeaders({ Authorization: `Bearer ${token}` });
  }

  createOrder(request: CreateOrderRequest): Observable<Order> {
    return this.http.post<Order>(this.BASE_URL, request, { headers: this.authHeaders() });
  }

  getOrderById(id: number): Observable<Order> {
    return this.http.get<Order>(`${this.BASE_URL}/${id}`, { headers: this.authHeaders() });
  }

  getMyOrders(): Observable<Order[]> {
    return this.http.get<Order[]>(`${this.BASE_URL}/my-orders`, { headers: this.authHeaders() });
  }

  getMySales(): Observable<Order[]> {
    return this.http.get<Order[]>(`${this.BASE_URL}/my-sales`, { headers: this.authHeaders() });
  }

  getOrdersByService(serviceId: number): Observable<Order[]> {
    return this.http.get<Order[]>(`${this.BASE_URL}/service/${serviceId}`, { headers: this.authHeaders() });
  }

  submitRequirements(orderId: number, requirementsAnswer: string): Observable<Order> {
    return this.http.patch<Order>(
      `${this.BASE_URL}/${orderId}/requirements`,
      { requirementsAnswer },
      { headers: this.authHeaders() }
    );
  }

  deliverOrder(orderId: number, deliveryMessage: string, deliveryFileUrls?: string): Observable<Order> {
    return this.http.patch<Order>(
      `${this.BASE_URL}/${orderId}/deliver`,
      { deliveryMessage, deliveryFileUrls },
      { headers: this.authHeaders() }
    );
  }

  requestRevision(orderId: number, revisionNotes: string): Observable<Order> {
    return this.http.patch<Order>(
      `${this.BASE_URL}/${orderId}/revision`,
      { revisionNotes },
      { headers: this.authHeaders() }
    );
  }

  completeOrder(orderId: number): Observable<Order> {
    return this.http.patch<Order>(
      `${this.BASE_URL}/${orderId}/complete`,
      {},
      { headers: this.authHeaders() }
    );
  }

  cancelOrder(orderId: number): Observable<Order> {
    return this.http.patch<Order>(
      `${this.BASE_URL}/${orderId}/cancel`,
      {},
      { headers: this.authHeaders() }
    );
  }

  disputeOrder(orderId: number): Observable<Order> {
    return this.http.patch<Order>(
      `${this.BASE_URL}/${orderId}/dispute`,
      {},
      { headers: this.authHeaders() }
    );
  }
}